In [1]:
import pandas as pd
import numpy as np

# Load the data set using pandas.read()

In [2]:
df=pd.read_csv(r"C:\Users\lenovo\Downloads\dataset\Retail_Transaction_Dataset.csv")

In [3]:
df

CustomerID ProductID  Quantity      Price   TransactionDate  \
0          109318         C         7  80.079844  12/26/2023 12:32   
1          993229         C         4  75.195229     8/5/2023 0:00   
2          579675         A         8  31.528816   3/11/2024 18:51   
3          799826         D         5  98.880218  10/27/2023 22:00   
4          121413         A         7  93.188512  12/22/2023 11:38   
...           ...       ...       ...        ...               ...   
99995      726461         A         2  56.078258   7/17/2023 16:59   
99996      328056         A         6  88.516406    5/30/2023 9:04   
99997      887304         B         4  72.385564    8/25/2023 7:59   
99998      326401         C         5  66.542239    2/5/2024 19:45   
99999      771566         C         5  38.087766    2/4/2024 11:53   

      PaymentMethod                                      StoreLocation  \
0              Cash            176 Andrew Cliffs\nBaileyfort, HI 93354   
1              Cash  11635 William Well Suite 809\nEast Kara, MT 19483   
2              Cash  910 Mendez Ville Suite 909\nPort Lauraland, MO...   
3            PayPal  87522 Sharon Corners Suite 500\nLake Tammy, MO...   
4              Cash   0070 Michelle Island Suite 143\nHoland, VA 80142   
...             ...                                                ...   
99995   Credit Card  3632 Darren Station Apt. 553\nEricaborough, RI...   
99996   Credit Card            821 Taylor Shoals\nEvansville, IL 70845   
99997   Credit Card       50653 Kara Lakes\nStephanieborough, RI 94492   
99998        PayPal  18756 Mcfarland Way Suite 866\nBarnettside, PR...   
99999    Debit Card                8046 Hull Drive\nPaulstad, GU 87218   

      ProductCategory  DiscountApplied(%)  TotalAmount  
0               Books           18.677100   455.862764  
1          Home Decor           14.121365   258.306546  
2               Books           15.943701   212.015651  
3               Books            6.686337   461.343769  
4         Electronics            4.030096   626.030484  
...               ...                 ...          ...  
99995        Clothing           18.345145    91.581240  
99996     Electronics            3.995541   509.878179  
99997        Clothing           17.423979   239.092472  
99998     Electronics           14.345018   284.983717  
99999      Home Decor            2.966058   184.790305  

[100000 rows x 10 columns]

# Observation
- Retail dataset has been loaded successfully using pandas read method(pd.read_csv())

# Understanding the structure of dataset

In [4]:
df.shape

(100000, 10)

## Observation
- The dataset contains 100000 rows and 10 columns representing complete details of a product like productId,payment mode,quantity,price,date and time of purchase etc by a customer.

# Previewing the first few rows

In [5]:
df.head()

CustomerID ProductID  Quantity      Price   TransactionDate PaymentMethod  \
0      109318         C         7  80.079844  12/26/2023 12:32          Cash   
1      993229         C         4  75.195229     8/5/2023 0:00          Cash   
2      579675         A         8  31.528816   3/11/2024 18:51          Cash   
3      799826         D         5  98.880218  10/27/2023 22:00        PayPal   
4      121413         A         7  93.188512  12/22/2023 11:38          Cash   

                                       StoreLocation ProductCategory  \
0            176 Andrew Cliffs\nBaileyfort, HI 93354           Books   
1  11635 William Well Suite 809\nEast Kara, MT 19483      Home Decor   
2  910 Mendez Ville Suite 909\nPort Lauraland, MO...           Books   
3  87522 Sharon Corners Suite 500\nLake Tammy, MO...           Books   
4   0070 Michelle Island Suite 143\nHoland, VA 80142     Electronics   

   DiscountApplied(%)  TotalAmount  
0           18.677100   455.862764  
1           14.121365   258.306546  
2           15.943701   212.015651  
3            6.686337   461.343769  
4            4.030096   626.030484

## Observation
- head() results the first 5 rows which help us understand that the file has loaded successfully.

# Dataset Structure and Datatypes

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 10 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   CustomerID          100000 non-null  int64  
 1   ProductID           100000 non-null  object 
 2   Quantity            100000 non-null  int64  
 3   Price               100000 non-null  float64
 4   TransactionDate     100000 non-null  object 
 5   PaymentMethod       100000 non-null  object 
 6   StoreLocation       100000 non-null  object 
 7   ProductCategory     100000 non-null  object 
 8   DiscountApplied(%)  100000 non-null  float64
 9   TotalAmount         100000 non-null  float64
dtypes: float64(3), int64(2), object(5)
memory usage: 7.6+ MB


# Observation
- The dataset contains 100000 rows and 10 columns.
- out of 10 columns:-
    - 2 columns contain integer type of values that is CustomerID and Quantity.
    - 5 columns contain categorical type of values including ProductID,PaymentMethod,TransactionDate etc.
    - 3 columns contain float type of values including TotalAmount,DiscountApplied,Price.

# Null Value count across all columns

In [8]:
df.isnull().sum()

CustomerID            0
ProductID             0
Quantity              0
Price                 0
TransactionDate       0
PaymentMethod         0
StoreLocation         0
ProductCategory       0
DiscountApplied(%)    0
TotalAmount           0
dtype: int64

## Observation
- This shows that there are no null or missing values.

# Statistical overview of numerical data.

In [9]:
df.describe()

CustomerID       Quantity          Price  DiscountApplied(%)  \
count  100000.000000  100000.000000  100000.000000       100000.000000   
mean   500463.982180       5.009290      55.067344           10.020155   
std    288460.917524       2.579808      25.971567            5.779534   
min        14.000000       1.000000      10.000430            0.000046   
25%    250693.750000       3.000000      32.549474            5.001013   
50%    499679.000000       5.000000      55.116789           10.030353   
75%    751104.750000       7.000000      77.456763           15.018367   
max    999997.000000       9.000000      99.999284           19.999585   

         TotalAmount  
count  100000.000000  
mean      248.334955  
std       184.554792  
min         8.274825  
25%        95.163418  
50%       200.368393  
75%       362.009980  
max       896.141242

## Observation
- This summary shows the statistical measures like count, mean, median, mode, min, max, standard deviation of all numerical columns

# Value Count Summary(Retail_Transaction_Details Dataset)

### 1.CustomerID
- This column contains 99999 unique id's, 1 id for each customer.

## 2. ProductID
- This column represents a ProductID for each product like Books, HomeDecors,Eletronics
- Here different products represent same ProductID.

## 3. Quantity
- This column contains the quantity of products purchased by customers.
- There are no missing values.
- This column completely contains integer type of values

## 4. Price
- This column contains the price of products sold by retail price product suppliers.
- This column represents float datatype of values.
  

## 5. TransactionDate
- This column contains date and time of purchase made by the customer.
- This column is of object datatype.

## 6. PaymentMethod
- This column represents the mode of payment made to purchase the product by the customer.
- Mode of payment is of different types like payment through Cash or Paypal or Debitcard or Creditcard

## 7. StoreLocation
- This column represents the address of the store where products are sold.

## 8. ProductCategory
- This column represents the different categories of products.
- Different products purchased by customers are Books, HomeDecor, Clothing, Electronics.

## 9. DiscountApplied(%)
- This column contains the discount percentage given for the products purchased by different customers.
- Percentage differs for different products.

## 10. TotalAmount
- This column represents the Total Amount calculated based on quantity, price and discount.
- This is the amount paid by the customer to purchase the product.

In [10]:
df["ProductID"].value_counts()

ProductID
C    25209
D    25042
B    24997
A    24752
Name: count, dtype: int64

In [11]:
df["Quantity"].value_counts()

Quantity
7    11236
6    11166
5    11146
9    11136
4    11135
8    11083
1    11064
2    11043
3    10991
Name: count, dtype: int64

In [12]:
df["PaymentMethod"].value_counts()

PaymentMethod
PayPal         25067
Credit Card    25034
Cash           25008
Debit Card     24891
Name: count, dtype: int64

In [13]:
df["ProductCategory"].value_counts()

ProductCategory
Clothing       25056
Books          25031
Electronics    24990
Home Decor     24923
Name: count, dtype: int64

In [14]:
df["TotalAmount"].value_counts()

TotalAmount
555.488605    2
101.718375    2
455.862764    1
398.737263    1
642.528301    1
             ..
109.481265    1
180.733181    1
638.254825    1
524.361666    1
184.790305    1
Name: count, Length: 99998, dtype: int64

# distribution of Key Fields

### Category column
- Payment, StoreLocation and product category are non numerical columns
- Product Category is well distributed across multiple categories such as Books, HomeDecor, Clothing, Electronics.
- Some categories like Books, Clothing appear more frequently compared to other categories.
- This suggest imbalance in sales across different categories.


### Target Variable
- Since this is retail dataset there is no paticular target variable.
- But according business point of view total amount can be treated as target variable for sales
- Even profit can be target variable if derived in later phase.

## Important Identifiers
- ProductID is not unique across categories that is same ID is repeated for different product categories.

In [15]:
df["CustomerID"].duplicated().sum()

4785

#### - Customer code repeats indicating multiple purchases are made by same customer

# Insights
### 1. There is a data quality issue as same ProductID is used for different Product Categories.
    - Product is not uniquely identified using ProductID.
    - if profits/total revenue or mostly sold products are calculated based on Product, it would produce wrong informtion and effect business.
### 2. CustomerID repeats many times, it shows same customer turned again to make purchases multiple times.
### 3. StoreLocation is unique and too detailed, if we extract city or state we can identify which region make more profits and expand expand business.
### 4. Based on statistics almost all category products  make equal revenue, Books make slightly more than clothing or electronics.

In [16]:
df.groupby("ProductCategory")["TotalAmount"].sum()

ProductCategory
Books          6.257837e+06
Clothing       6.205502e+06
Electronics    6.196735e+06
Home Decor     6.173421e+06
Name: TotalAmount, dtype: float64

In [17]:
df.groupby("ProductCategory")["TotalAmount"].mean()

ProductCategory
Books          250.003477
Clothing       247.665324
Electronics    247.968578
Home Decor     247.699770
Name: TotalAmount, dtype: float64

In [18]:
df[df.duplicated(subset=['CustomerID'], keep=False)]

CustomerID ProductID  Quantity      Price   TransactionDate  \
4          121413         A         7  93.188512  12/22/2023 11:38   
19         818911         B         5  77.090575    4/6/2024 18:59   
27         418277         A         9  27.450194  11/29/2023 11:40   
32         628270         A         5  61.432803     6/4/2023 0:37   
41         147124         C         2  79.673863   5/25/2023 10:49   
...           ...       ...       ...        ...               ...   
99931      444530         D         5  89.717996   5/23/2023 13:12   
99938      730233         A         4  40.081843     2/1/2024 3:15   
99954      666391         D         6  30.980492    8/25/2023 1:51   
99974      655723         B         2  49.992310    4/24/2024 8:24   
99980      488114         C         7  65.958432    2/3/2024 10:57   

      PaymentMethod                                      StoreLocation  \
4              Cash   0070 Michelle Island Suite 143\nHoland, VA 80142   
19      Credit Card  44623 Elizabeth Throughway Apt. 440\nAmberview...   
27      Credit Card  71524 Margaret Dam\nNorth Rhondaborough, VA 49112   
32       Debit Card  6472 Jones Courts Apt. 861\nWest Dawnport, KS ...   
41             Cash                           USNS David\nFPO AE 12953   
...             ...                                                ...   
99931    Debit Card  6552 Wheeler Isle Suite 644\nCrystalside, ID 5...   
99938    Debit Card      05350 Peck Cliff Apt. 080\nFreyside, KY 92968   
99954          Cash    9445 Michael Track\nNorth Abigailside, AK 94088   
99974          Cash  57588 Kevin Trafficway\nLake Timothyshire, IL ...   
99980        PayPal            051 Hale Divide\nDavidborough, PR 86800   

      ProductCategory  DiscountApplied(%)  TotalAmount  
4         Electronics            4.030096   626.030484  
19              Books           19.252574   311.243274  
27           Clothing            4.398935   236.184096  
32              Books            6.323548   287.740350  
41           Clothing           12.913782   138.769908  
...               ...                 ...          ...  
99931      Home Decor           10.918878   399.608987  
99938           Books            3.533413   154.662345  
99954           Books            9.047247   169.065663  
99974           Books           18.753029    81.234475  
99980           Books            5.377907   436.878742  

[9406 rows x 10 columns]

# Outliers

In [19]:
Q1=df["TotalAmount"].quantile(0.25)
Q1

95.1634184575

In [20]:
Q2=df["TotalAmount"].quantile(0.50)
Q2

200.3683931

In [21]:
Q3=df["TotalAmount"].quantile(0.75)
Q3

362.00997989999996

In [22]:
IQR=Q3-Q1
IQR

266.84656144249993

In [23]:
lower_boundary=Q1-(1.5*IQR)
lower_boundary

-305.10642370624987

In [24]:
upper_boundary=Q3+(1.5*IQR)
upper_boundary

762.2798220637499

In [25]:
df[(df["TotalAmount"]<lower_boundary)|(df["TotalAmount"]>upper_boundary)]

CustomerID ProductID  Quantity      Price   TransactionDate  \
107        142499         B         9  88.778165     5/8/2023 0:23   
464        493138         D         9  99.999284   9/27/2023 23:40   
496        587590         D         9  88.377706   8/25/2023 23:40   
662        850461         B         8  99.566848    1/7/2024 23:04   
663        740108         C         9  99.981646   9/16/2023 23:36   
...           ...       ...       ...        ...               ...   
99569      891249         D         9  94.731617     2/4/2024 2:45   
99607      424560         B         9  92.972276   11/2/2023 16:00   
99732      243629         C         9  89.621820   10/5/2023 10:47   
99762      159085         B         9  95.568751    4/26/2024 0:19   
99975      507259         A         9  98.097683  11/18/2023 15:08   

      PaymentMethod                                      StoreLocation  \
107     Credit Card              409 Hampton Hill\nLake Gail, MA 03704   
464     Credit Card            766 Bryant Lock\nNorth Ashley, MP 73093   
496          PayPal  8711 Jonathon Vista Apt. 368\nLoriburgh, WI 89346   
662      Debit Card             4193 Kevin Knolls\nEvanstown, MA 73699   
663          PayPal  0721 Eric Points Suite 904\nMarioborough, MO 0...   
...             ...                                                ...   
99569        PayPal  5532 Nancy Causeway Suite 281\nFelicialand, RI...   
99607          Cash  648 Burnett Hollow Suite 062\nPort Chad, PA 98954   
99732   Credit Card  501 Jessica Passage Suite 048\nJasminebury, MS...   
99762          Cash  599 Ashley Route Apt. 451\nNorth Brittneyshire...   
99975    Debit Card     965 Meza Ville Apt. 863\nSalazarside, PW 27127   

      ProductCategory  DiscountApplied(%)  TotalAmount  
107       Electronics            0.888264   791.906223  
464          Clothing            2.093143   881.155400  
496             Books            3.986390   763.691639  
662       Electronics            1.129157   787.540655  
663       Electronics           14.297624   771.179815  
...               ...                 ...          ...  
99569     Electronics            0.995650   844.095797  
99607      Home Decor            4.946829   795.357871  
99732      Home Decor            2.753572   784.386168  
99762           Books            8.647966   785.735978  
99975      Home Decor            9.133472   802.241629  

[853 rows x 10 columns]

In [26]:
df[(df["TotalAmount"]<lower_boundary)]

Empty DataFrame
Columns: [CustomerID, ProductID, Quantity, Price, TransactionDate, PaymentMethod, StoreLocation, ProductCategory, DiscountApplied(%), TotalAmount]
Index: []

In [27]:
df[(df["TotalAmount"]>upper_boundary)]

CustomerID ProductID  Quantity      Price   TransactionDate  \
107        142499         B         9  88.778165     5/8/2023 0:23   
464        493138         D         9  99.999284   9/27/2023 23:40   
496        587590         D         9  88.377706   8/25/2023 23:40   
662        850461         B         8  99.566848    1/7/2024 23:04   
663        740108         C         9  99.981646   9/16/2023 23:36   
...           ...       ...       ...        ...               ...   
99569      891249         D         9  94.731617     2/4/2024 2:45   
99607      424560         B         9  92.972276   11/2/2023 16:00   
99732      243629         C         9  89.621820   10/5/2023 10:47   
99762      159085         B         9  95.568751    4/26/2024 0:19   
99975      507259         A         9  98.097683  11/18/2023 15:08   

      PaymentMethod                                      StoreLocation  \
107     Credit Card              409 Hampton Hill\nLake Gail, MA 03704   
464     Credit Card            766 Bryant Lock\nNorth Ashley, MP 73093   
496          PayPal  8711 Jonathon Vista Apt. 368\nLoriburgh, WI 89346   
662      Debit Card             4193 Kevin Knolls\nEvanstown, MA 73699   
663          PayPal  0721 Eric Points Suite 904\nMarioborough, MO 0...   
...             ...                                                ...   
99569        PayPal  5532 Nancy Causeway Suite 281\nFelicialand, RI...   
99607          Cash  648 Burnett Hollow Suite 062\nPort Chad, PA 98954   
99732   Credit Card  501 Jessica Passage Suite 048\nJasminebury, MS...   
99762          Cash  599 Ashley Route Apt. 451\nNorth Brittneyshire...   
99975    Debit Card     965 Meza Ville Apt. 863\nSalazarside, PW 27127   

      ProductCategory  DiscountApplied(%)  TotalAmount  
107       Electronics            0.888264   791.906223  
464          Clothing            2.093143   881.155400  
496             Books            3.986390   763.691639  
662       Electronics            1.129157   787.540655  
663       Electronics           14.297624   771.179815  
...               ...                 ...          ...  
99569     Electronics            0.995650   844.095797  
99607      Home Decor            4.946829   795.357871  
99732      Home Decor            2.753572   784.386168  
99762           Books            8.647966   785.735978  
99975      Home Decor            9.133472   802.241629  

[853 rows x 10 columns]

In [41]:
def Outliers(df, column):
    Q1 = df[column].quantile(0.25)   
    Q3 = df[column].quantile(0.75)   
    IQR = Q3 - Q1                    
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers = df[(df[column] < lower_bound) | (df[column] > upper_bound)]
    
    print(f"Column: {column}")
    print(f"Lower Bound: {lower_bound}, Upper Bound: {upper_bound}")
    print(f"Total Outliers: {outliers}")
    
    return outliers


In [42]:
find_outliers(df,"Price")

Column: Price
Lower Bound: -34.8114600025, Upper Bound: 144.81769747750002
Total Outliers: 0


Empty DataFrame
Columns: [CustomerID, ProductID, Quantity, Price, TransactionDate, PaymentMethod, StoreLocation, ProductCategory, DiscountApplied(%), TotalAmount]
Index: []

In [33]:
find_outliers(df,"DiscountApplied(%)")

Column: DiscountApplied(%)
Lower Bound: -10.025019071875002, Upper Bound: 30.044398599125003
Total Outliers: 0


Empty DataFrame
Columns: [CustomerID, ProductID, Quantity, Price, TransactionDate, PaymentMethod, StoreLocation, ProductCategory, DiscountApplied(%), TotalAmount]
Index: []

In [34]:
find_outliers(df,"CustomerID")

Column: CustomerID
Lower Bound: -499922.75, Upper Bound: 1501721.25
Total Outliers: 0


Empty DataFrame
Columns: [CustomerID, ProductID, Quantity, Price, TransactionDate, PaymentMethod, StoreLocation, ProductCategory, DiscountApplied(%), TotalAmount]
Index: []

In [35]:
find_outliers(df,"Quantity")

Column: Quantity
Lower Bound: -3.0, Upper Bound: 13.0
Total Outliers: 0


Empty DataFrame
Columns: [CustomerID, ProductID, Quantity, Price, TransactionDate, PaymentMethod, StoreLocation, ProductCategory, DiscountApplied(%), TotalAmount]
Index: []

In [43]:
find_outliers(df,"TotalAmount")

Column: TotalAmount
Lower Bound: -305.10642370624987, Upper Bound: 762.2798220637499
Total Outliers: 853


CustomerID ProductID  Quantity      Price   TransactionDate  \
107        142499         B         9  88.778165     5/8/2023 0:23   
464        493138         D         9  99.999284   9/27/2023 23:40   
496        587590         D         9  88.377706   8/25/2023 23:40   
662        850461         B         8  99.566848    1/7/2024 23:04   
663        740108         C         9  99.981646   9/16/2023 23:36   
...           ...       ...       ...        ...               ...   
99569      891249         D         9  94.731617     2/4/2024 2:45   
99607      424560         B         9  92.972276   11/2/2023 16:00   
99732      243629         C         9  89.621820   10/5/2023 10:47   
99762      159085         B         9  95.568751    4/26/2024 0:19   
99975      507259         A         9  98.097683  11/18/2023 15:08   

      PaymentMethod                                      StoreLocation  \
107     Credit Card              409 Hampton Hill\nLake Gail, MA 03704   
464     Credit Card            766 Bryant Lock\nNorth Ashley, MP 73093   
496          PayPal  8711 Jonathon Vista Apt. 368\nLoriburgh, WI 89346   
662      Debit Card             4193 Kevin Knolls\nEvanstown, MA 73699   
663          PayPal  0721 Eric Points Suite 904\nMarioborough, MO 0...   
...             ...                                                ...   
99569        PayPal  5532 Nancy Causeway Suite 281\nFelicialand, RI...   
99607          Cash  648 Burnett Hollow Suite 062\nPort Chad, PA 98954   
99732   Credit Card  501 Jessica Passage Suite 048\nJasminebury, MS...   
99762          Cash  599 Ashley Route Apt. 451\nNorth Brittneyshire...   
99975    Debit Card     965 Meza Ville Apt. 863\nSalazarside, PW 27127   

      ProductCategory  DiscountApplied(%)  TotalAmount  
107       Electronics            0.888264   791.906223  
464          Clothing            2.093143   881.155400  
496             Books            3.986390   763.691639  
662       Electronics            1.129157   787.540655  
663       Electronics           14.297624   771.179815  
...               ...                 ...          ...  
99569     Electronics            0.995650   844.095797  
99607      Home Decor            4.946829   795.357871  
99732      Home Decor            2.753572   784.386168  
99762           Books            8.647966   785.735978  
99975      Home Decor            9.133472   802.241629  

[853 rows x 10 columns]

### Observation
- It is observed that TotalAmount have outliers but those are just larger revenue value not errors.

In [45]:

summary = df.groupby("ProductCategory").agg(
    avg_price_per_item=("Price", "mean"),
    avg_quantity=("Quantity", "mean"),
    avg_total_amount=("TotalAmount", "mean"),
    total_revenue=("TotalAmount", "sum"),
    transaction_count=("TotalAmount", "count")
).reset_index()

print(summary)


  ProductCategory  avg_price_per_item  avg_quantity  avg_total_amount  \
0           Books           55.101589      5.035636        250.003477   
1        Clothing           55.127468      4.990581        247.665324   
2     Electronics           54.958181      5.015886        247.968578   
3      Home Decor           55.081963      4.995025        247.699770   

   total_revenue  transaction_count  
0   6.257837e+06              25031  
1   6.205502e+06              25056  
2   6.196735e+06              24990  
3   6.173421e+06              24923  
